katz/src/main/kotlin/katz/typeclasses/Foldable.kt

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) 2017 The Katz Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package katz
+
+import katz.Eval.EvalFactory.always
+
+/**
+ * Data structures that can be folded to a summary value.
+ *
+ * In the case of a collection (such as `List` or `Set`), these methods will fold together (combine) the values
+ * contained in the collection to produce a single result. Most collection types have `foldLeft` methods, which will
+ * usually be used by the associated `Foldable[_]` instance.
+ *
+ * Foldable[F] is implemented in terms of two basic methods:
+ *
+ *  - `foldLeft(fa, b)(f)` eagerly folds `fa` from left-to-right.
+ *  - `foldRight(fa, b)(f)` lazily folds `fa` from right-to-left.
+ *
+ * Beyond these it provides many other useful methods related to folding over F[A] values.
+ */
+interface Foldable<F> {
```

Comment:
Since we need also `Traverse` for `sequence` and other useful combinators we should attempt to include that as part of this PR if you think it's appropriate, otherwise a separate PR once this is integrated is fine.